File: one_away.js
#!/usr/bin/env node

"use strict";

function one_edit_away(s1, s2) {
  if (s1.length == s2.length) {
    return one_edit_replace(s1, s2);
  } else if (s1.length + 1 == s2.length) {
    return one_edit_insert(s1, s2);
  } else if (s2.length + 1 == s1.length) {
    return one_edit_insert(s2, s1);
  } else {
    return false;
  }
}

function one_edit_replace(s1, s2) {
  var found_difference = false;
  for (let i = 0; i < s1.length; i++) {
    if (s1[i] !== s2[i]) {
      if (found_difference) {
        return false;
      }
      found_difference = true;
    }
  }
  return true;
}

function one_edit_insert(s1, s2) {
  var x1 = 0;
  var x2 = 0;
  while (x1 < s1.length && x2 < s2.length) {
    if (s1[x1] != s2[x2]) {
      if (x1 !== x2) {
        return false;
      }
    } else {
      x1 += 1;
    }
    x2 += 1;
  }
  return true;
}

console.log(one_edit_away('pale', 'ple'));   // true
console.log(one_edit_away('pale', 'pales')); // true
console.log(one_edit_away('pale', 'bale'));  // true
console.log(one_edit_away('pale', 'base'));  // false
console.log(one_edit_away('pale', 'bae'));   // false
console.log(one_edit_away('pale', 'le'));    // false

File: one_away.py
#!/usr/bin/env python

def one_edit_away(s1, s2):
  if len(s1) == len(s2):
    return one_edit_replace(s1, s2)
  elif len(s1) + 1 == len(s2):
    return one_edit_insert(s1, s2)
  elif len(s2) + 1 == len(s1):
    return one_edit_insert(s2, s1)
  else:
    return False

def one_edit_replace(s1, s2):
  found_difference = False
  for i in range(0, len(s1)):
    if s1[i] != s2[i]:
      if found_difference:
        return False
      found_difference = True
  return True

def one_edit_insert(s1, s2):
  x1 = x2 = 0
  while x1 < len(s1) and x2 < len(s2):
    if s1[x1] != s2[x2]:
      if x1 != x2:
        return False
    else:
      x1 += 1
    x2 += 1
  return True

print one_edit_away('pale', 'ple')   # True
print one_edit_away('pale', 'pales') # True
print one_edit_away('pale', 'bale')  # True
print one_edit_away('pale', 'base')  # False
print one_edit_away('pale', 'bae')   # False
print one_edit_away('pale', 'le')    # False
